nflog/nflog.go

```diff
@@ -289,48 +216,79 @@ func marshalMeshEntry(e *pb.MeshEntry) ([]byte, error) {
 	return buf.Bytes(), nil
 }
 
+// Options configures a new Log implementation.
+type Options struct {
+	SnapshotReader io.Reader
+	SnapshotFile   string
+
+	Retention time.Duration
+
+	Logger  log.Logger
+	Metrics prometheus.Registerer
+}
+
+func (o *Options) validate() error {
+	if o.SnapshotFile != "" && o.SnapshotReader != nil {
+		return errors.New("only one of SnapshotFile and SnapshotReader must be set")
+	}
+
+	return nil
+}
+
 // New creates a new notification log based on the provided options.
 // The snapshot is loaded into the Log if it is set.
-func New(opts ...Option) (*Log, error) {
+func New(o Options) (*Log, error) {
+	if err := o.validate(); err != nil {
+		return nil, err
+	}
+
+	if o.SnapshotFile != "" {
+		if r, err := os.Open(o.SnapshotFile); err != nil {
+			if !os.IsNotExist(err) {
+				return nil, err
+			}
+		} else {
+			o.SnapshotReader = r
+			defer r.Close()
+		}
+	}
+
 	l := &Log{
+		clock:     clock.New(),
+		retention: o.Retention,
 		logger:    log.NewNopLogger(),
-		now:       utcNow,
 		st:        state{},
 		broadcast: func([]byte) {},
-	}
-	for _, o := range opts {
-		if err := o(l); err != nil {
-			return nil, err
-		}
-	}
-	if l.metrics == nil {
-		l.metrics = newMetrics(nil)
+		metrics:   newMetrics(o.Metrics),
 	}
 
-	if l.snapf != "" {
-		if f, err := os.Open(l.snapf); !os.IsNotExist(err) {
-			if err != nil {
-				return l, err
-			}
-			defer f.Close()
+	if o.Logger != nil {
+		l.logger = o.Logger
+	}
 
-			if err := l.loadSnapshot(f); err != nil {
-				return l, err
-			}
+	if o.SnapshotReader != nil {
+		if err := l.loadSnapshot(o.SnapshotReader); err != nil {
+			return l, err
 		}
 	}
 
-	go l.run()
-
 	return l, nil
 }
 
-// run periodic background maintenance.
-func (l *Log) run() {
-	if l.runInterval == 0 || l.stopc == nil {
+func (l *Log) now() time.Time {
+	return l.clock.Now()
+}
+
+// Maintenance garbage collects the notification log state at the given interval. If the snapshot
+// file is set, a snapshot is written to it afterwards.
+// Terminates on receiving from stopc.
+// If not nil, the last argument is an override for what to do as part of the maintenance - for advanced usage.
+func (l *Log) Maintenance(interval time.Duration, snapf string, stopc <-chan struct{}, override MaintenanceFunc) {
+	if interval == 0 || stopc == nil {
+		level.Error(l.logger).Log("msg", "interval or stop signal are missing - not running maintenance")
```

Comment:
This is new - it kind of annoyed me that we'd return from this function on silences when you misconfigured the maintenance but fail silently. I don't think we should do this -- ideally, return an error but I settled with a log line to keep the diff sane.

Reply:
Works for me. I'd suggest that we check the validity of the maintenance interval in cmd/alertmanager/main.go in a later PR. As of today, a negative maintenance interval triggers a panic...